pkg/cmd/pulumi/util_remote.go

```diff
@@ -0,0 +1,251 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/result"
+)
+
+// This is a variable instead of a constant so it can be set in certain builds of the CLI that do not
+// support remote deployments.
+var disableRemote bool
+
+// remoteSupported returns true if the CLI supports remote deployments.
+func remoteSupported() bool {
+	return !disableRemote && hasExperimentalCommands()
+}
+
+// parseEnv parses a `--remote-env` flag value for `--remote`. A value should be of the form
+// "NAME=value" for regular envvars or "NAME#=value" for envvars with a secret value.
+func parseEnv(input string) (string, string, bool, error) {
+	pair := strings.SplitN(input, "=", 2)
+	if len(pair) != 2 {
+		return "", "", false, fmt.Errorf("error parsing %q", input)
+	}
+	name, value := pair[0], pair[1]
+
+	var secret bool
+	if last := len(name) - 1; last >= 0 && name[last] == '#' {
+		secret = true
+		name = name[:last]
+	}
+
+	if name == "" {
+		return "", "", false, fmt.Errorf("expected non-empty environment name for %q", input)
+	}
+
+	return name, value, secret, nil
+}
+
+// validateUnsupportedRemoteFlags returns an error if any unsupported flags are set when --remote is set.
+func validateUnsupportedRemoteFlags(
+	expectNop bool,
+	configArray []string,
+	configPath bool,
+	client string,
+	jsonDisplay bool,
+	policyPackPaths []string,
+	policyPackConfigPaths []string,
+	refresh string,
+	showConfig bool,
+	showReplacementSteps bool,
+	showSames bool,
+	showReads bool,
+	suppressOutputs bool,
+	secretsProvider string,
+	targets *[]string,
+	replaces []string,
+	targetReplaces []string,
+	targetDependents bool,
+	planFilePath string,
+	stackConfigFile string,
+) error {
+
+	if expectNop {
+		return errors.New("--expect-no-changes is not supported with --remote")
+	}
+	if len(configArray) > 0 {
+		return errors.New("--config is not supported with --remote")
+	}
+	if configPath {
+		return errors.New("--config-path is not supported with --remote")
+	}
+	if client != "" {
+		return errors.New("--client is not supported with --remote")
+	}
+	// We should be able to make --json work, but it doesn't work currently.
+	if jsonDisplay {
+		return errors.New("--json is not supported with --remote")
```

Comment:
Automation API operations rely on `--json` (for example in dotnet) to return typed results, wouldn't it be required for `--json` to work with`--remote`?

Reply:
`--json` isn't needed by Automation API in .NET (or the other languages) for preview, update, refresh, or destroy.

That said, I do want `--json` to work when `--remote` is passed, there's just some extraneous output and a panic that needs to be fixed before it can be used, so I'm just explicitly disabling it for now. I'm planning to fix this after these initial changes land.